lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -599,13 +626,73 @@ LifetimeChecker::LifetimeChecker(const DIMemoryObjectInfo &TheMemory,
     // FIXME: critical edges?
     auto *bb = I->getParent();
     getBlockInfo(bb).markStoreToSelf();
+
+    // Track StoresToSelf as liveness pre-computation candidates.
+    recordInstForLocalLivenessCheck(I);
+  }
+
+  // Track Destroys as liveness pre-computation candidates.
+  for (auto Release : Destroys) {
+    recordInstForLocalLivenessCheck(Release);
   }
 
   // It isn't really a use, but we account for the mark_uninitialized or
   // project_box as a use so we see it in our dataflow walks.
   auto &MemBBInfo = getBlockInfo(TheMemory.getParentBlock());
   MemBBInfo.HasNonLoadUse = true;
 
+  // Track the memory instruction for pre-computation
+  recordInstForLocalLivenessCheck(TheMemory.getUninitializedValue());
+  // TODO: ^ is this right? TheMemory doesn't give direct access to its Inst
+
+  // We've collected the instructions & blocks we expect to be queried for
+  // liveness, so now we can pre-compute the intra-block memory availability
+  // at these instructions to speed up the anticipated future queries.
+  for (auto InstBB : BlocksOfNote) {
+    auto &BBInfo = getBlockInfo(InstBB);
+
+    // If there aren't any non-load uses, there's nothing to pre-compute.
+    if (!BBInfo.HasNonLoadUse)
+      continue;
+
+    // Already computed the intra-block cache, so bail.
+    if (BBInfo.InstToAvailability.has_value())
+      continue;
```

Comment:
i didn't realize the same `LiveOutBlockState` will be reused when processing multiple distinct mark_uninitialized instructions within the same block, so i'm now a bit less sure if this approach is correct. we don't want to accidentally consult a stale cache... will have to think a bit more whether something along these lines has the potential to address this particular performance problem.

Reply:
seems i was likely mistaken about the behavior here – the map from block to state is reused, but the state itself is reinitialized for each memory object being processed (makes sense, otherwise there'd be serious issues with DI presumably).